[Bug]: Check Files Locked not correctly export to the remote server

### Apache Hop version?

2.11

### Java version?

22

### Operating system

Windows

### What happened?

Hi, 

There is a bug with the generated xml file when I want to execute a workflow with the node : "Check Files Locked" on a remote server from hop gui.

<img width="491" height="276" alt="Image" src="https://github.com/user-attachments/assets/6d7cc370-6533-4878-a83a-c518f8c3a68d" />

Thx

The result is the java error :

org.apache.hop.core.exception.HopXmlException: Unable to load the workflow from XML file [zip:file:///C:/Users/ADMINI~1.DIS/AppData/Local/Temp/1/export_41adbff1-bf89-442e-bb5e-1e62eb36f70c.zip!transfert_fichiers_solid.hwf]Unable to load workflow info from XML nodeUnable to load workflow hop info from XML nodeCannot invoke "String.equalsIgnoreCase(String)" because the return value of "org.apache.hop.workflow.action.ActionMeta.getName()" is null	at org.apache.hop.workflow.WorkflowMeta.loadXml(WorkflowMeta.java:452)	at org.apache.hop.workflow.WorkflowMeta.<init>(WorkflowMeta.java:433)	at org.apache.hop.www.RegisterPackageServlet.generateBody(RegisterPackageServlet.java:106)	at org.apache.hop.www.BodyHttpServlet.doGet(BodyHttpServlet.java:65)	at org.apache.hop.www.BaseHttpServlet.doPost(BaseHttpServlet.java:114)	at javax.servlet.http.HttpServlet.service(HttpServlet.java:707)	at org.apache.hop.www.BaseHttpServlet.service(BaseHttpServlet.java:102)	at javax.servlet.http.HttpServlet.service(HttpServlet.java:790)	at org.eclipse.jetty.servlet.ServletHolder.handle(ServletHolder.java:799)	at org.eclipse.jetty.servlet.ServletHandler.doHandle(ServletHandler.java:550)	at org.eclipse.jetty.server.handler.ScopedHandler.nextHandle(ScopedHandler.java:233)	at org.eclipse.jetty.server.session.SessionHandler.doHandle(SessionHandler.java:1624)	at org.eclipse.jetty.server.handler.ScopedHandler.nextHandle(ScopedHandler.java:233)	at org.eclipse.jetty.server.handler.ContextHandler.doHandle(ContextHandler.java:1435)	at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:188)	at org.eclipse.jetty.servlet.ServletHandler.doScope(ServletHandler.java:501)	at org.eclipse.jetty.server.session.SessionHandler.doScope(SessionHandler.java:1594)	at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:186)	at org.eclipse.jetty.server.handler.ContextHandler.doScope(ContextHandler.java:1350)	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:141)	at org.eclipse.jetty.server.handler.ContextHandlerCollection.handle(ContextHandlerCollection.java:234)	at org.eclipse.jetty.server.handler.HandlerList.handle(HandlerList.java:59)	at org.eclipse.jetty.security.SecurityHandler.handle(SecurityHandler.java:567)	at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:127)	at org.eclipse.jetty.server.Server.handle(Server.java:516)	at org.eclipse.jetty.server.HttpChannel.lambda$handle$1(HttpChannel.java:388)	at org.eclipse.jetty.server.HttpChannel.dispatch(HttpChannel.java:633)	at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:380)	at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:277)	at org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:311)	at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:105)	at org.eclipse.jetty.io.ChannelEndPoint$1.run(ChannelEndPoint.java:104)	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.runTask(EatWhatYouKill.java:336)	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.doProduce(EatWhatYouKill.java:313)	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.tryProduce(EatWhatYouKill.java:171)	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.run(EatWhatYouKill.java:129)	at org.eclipse.jetty.util.thread.ReservedThreadExecutor$ReservedThread.run(ReservedThreadExecutor.java:383)	at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:882)	at org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.run(QueuedThreadPool.java:1036)	at java.base/java.lang.Thread.run(Thread.java:1570)Caused by: org.apache.hop.core.exception.HopXmlException: Unable to load workflow info from XML nodeUnable to load workflow hop info from XML nodeCannot invoke "String.equalsIgnoreCase(String)" because the return value of "org.apache.hop.workflow.action.ActionMeta.getName()" is null	at org.apache.hop.workflow.WorkflowMeta.loadXml(WorkflowMeta.java:609)	at org.apache.hop.workflow.WorkflowMeta.loadXml(WorkflowMeta.java:445)	... 39 moreCaused by: org.apache.hop.core.exception.HopXmlException: Unable to load workflow hop info from XML nodeCannot invoke "String.equalsIgnoreCase(String)" because the return value of "org.apache.hop.workflow.action.ActionMeta.getName()" is null	at org.apache.hop.workflow.WorkflowHopMeta.<init>(WorkflowHopMeta.java:118)	at org.apache.hop.workflow.WorkflowMeta.loadXml(WorkflowMeta.java:583)	... 40 moreCaused by: java.lang.NullPointerException: Cannot invoke "String.equalsIgnoreCase(String)" because the return value of "org.apache.hop.workflow.action.ActionMeta.getName()" is null	at org.apache.hop.workflow.WorkflowMeta.findAction(WorkflowMeta.java:800)	at org.apache.hop.workflow.WorkflowHopMeta.<init>(WorkflowHopMeta.java:112)	... 41 more

### Issue Priority

Priority: 3

### Issue Component

Component: Hop Server

--- a/engine/src/main/java/org/apache/hop/workflow/action/ActionBase.java
+++ b/engine/src/main/java/org/apache/hop/workflow/action/ActionBase.java
@@ -170,10 +170,14 @@ public abstract class ActionBase
   /**
    * Copy constructor variant
    *
+   * <p>The plugin id is copied as well: resource export clones every action, and a clone without a
+   * plugin id is written to the export without its {@code <type>} element, which then fails to load
+   * ("Unable to load workflow info from XML node"). See issue #5644.
+   *
    * @param b The action base to copy
    */
   protected ActionBase(ActionBase b) {
-    this(b.name, b.description);
+    this(b.name, b.description, b.pluginId);
   }
 
   /**

--- a/integration-tests/hop_server/0010-check-files-locked.hwf
+++ b/integration-tests/hop_server/0010-check-files-locked.hwf
@@ -1,0 +1,127 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!--
+
+Licensed to the Apache Software Foundation (ASF) under one or more
+contributor license agreements.  See the NOTICE file distributed with
+this work for additional information regarding copyright ownership.
+The ASF licenses this file to You under the Apache License, Version 2.0
+(the "License"); you may not use this file except in compliance with
+the License.  You may obtain a copy of the License at
+
+      http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+
+-->
+<workflow>
+  <name>0010-check-files-locked</name>
+  <name_sync_with_filename>Y</name_sync_with_filename>
+  <description>Contains a Check Files Locked action and is executed remotely by the parent workflow,
+    so it is serialized through the resource export. Resource export clones every action; if an
+    action's copy constructor does not carry over its name and plugin id, the exported XML holds an
+    empty name and no type element, and loading it fails on the server (see issue #5644).
+
+    The file to check is created by this workflow rather than taken from the project, so that it
+    exists on the file system of whichever host runs the export.</description>
+  <extended_description/>
+  <workflow_version/>
+  <created_user>-</created_user>
+  <created_date>2024/01/01 00:00:00.000</created_date>
+  <modified_user>-</modified_user>
+  <modified_date>2024/01/01 00:00:00.000</modified_date>
+  <parameters>
+    </parameters>
+  <actions>
+    <action>
+      <name>Start</name>
+      <description/>
+      <type>SPECIAL</type>
+      <attributes/>
+      <DayOfMonth>1</DayOfMonth>
+      <hour>12</hour>
+      <intervalMinutes>60</intervalMinutes>
+      <intervalSeconds>0</intervalSeconds>
+      <minutes>0</minutes>
+      <repeat>N</repeat>
+      <schedulerType>0</schedulerType>
+      <weekDay>1</weekDay>
+      <parallel>N</parallel>
+      <xloc>50</xloc>
+      <yloc>50</yloc>
+      <attributes_hac/>
+    </action>
+    <action>
+      <name>Create file to check</name>
+      <description/>
+      <type>CREATE_FILE</type>
+      <attributes/>
+      <filename>${java.io.tmpdir}/0010-not-locked.txt</filename>
+      <fail_if_file_exists>N</fail_if_file_exists>
+      <add_filename_result>N</add_filename_result>
+      <parallel>N</parallel>
+      <xloc>250</xloc>
+      <yloc>50</yloc>
+      <attributes_hac/>
+    </action>
+    <action>
+      <name>Check files locked</name>
+      <description/>
+      <type>CHECK_FILES_LOCKED</type>
+      <attributes/>
+      <arg_from_previous>N</arg_from_previous>
+      <fields>
+        <field>
+          <name>${java.io.tmpdir}/0010-not-locked.txt</name>
+          <filemask/>
+        </field>
+      </fields>
+      <include_subfolders>N</include_subfolders>
+      <parallel>N</parallel>
+      <xloc>450</xloc>
+      <yloc>50</yloc>
+      <attributes_hac/>
+    </action>
+    <action>
+      <name>Delete file to check</name>
+      <description/>
+      <type>DELETE_FILE</type>
+      <attributes/>
+      <filename>${java.io.tmpdir}/0010-not-locked.txt</filename>
+      <fail_if_file_not_exists>Y</fail_if_file_not_exists>
+      <parallel>N</parallel>
+      <xloc>650</xloc>
+      <yloc>50</yloc>
+      <attributes_hac/>
+    </action>
+  </actions>
+  <hops>
+    <hop>
+      <from>Start</from>
+      <to>Create file to check</to>
+      <enabled>Y</enabled>
+      <evaluation>Y</evaluation>
+      <unconditional>Y</unconditional>
+    </hop>
+    <hop>
+      <from>Create file to check</from>
+      <to>Check files locked</to>
+      <enabled>Y</enabled>
+      <evaluation>Y</evaluation>
+      <unconditional>N</unconditional>
+    </hop>
+    <hop>
+      <from>Check files locked</from>
+      <to>Delete file to check</to>
+      <enabled>Y</enabled>
+      <evaluation>Y</evaluation>
+      <unconditional>N</unconditional>
+    </hop>
+  </hops>
+  <notepads>
+  </notepads>
+  <attributes/>
+</workflow>

--- a/integration-tests/hop_server/main-0010-test-check-files-locked-export.hwf
+++ b/integration-tests/hop_server/main-0010-test-check-files-locked-export.hwf
@@ -1,0 +1,95 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!--
+
+Licensed to the Apache Software Foundation (ASF) under one or more
+contributor license agreements.  See the NOTICE file distributed with
+this work for additional information regarding copyright ownership.
+The ASF licenses this file to You under the Apache License, Version 2.0
+(the "License"); you may not use this file except in compliance with
+the License.  You may obtain a copy of the License at
+
+      http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+
+-->
+<workflow>
+  <name>main-0010-test-check-files-locked-export</name>
+  <name_sync_with_filename>Y</name_sync_with_filename>
+  <description>Regression test for issue #5644: a workflow containing a Check Files Locked action must
+    survive the resource export to a remote Hop Server. The inner workflow is executed remotely with
+    the "remote" run configuration (export_resources=true). Resource export clones every action, so
+    an action whose copy constructor drops its name is exported with an empty name; loading that XML
+    then fails with "Cannot invoke String.equalsIgnoreCase(String) because the return value of
+    ActionMeta.getName() is null" and this workflow fails.</description>
+  <extended_description/>
+  <workflow_version/>
+  <created_user>-</created_user>
+  <created_date>2024/01/01 00:00:00.000</created_date>
+  <modified_user>-</modified_user>
+  <modified_date>2024/01/01 00:00:00.000</modified_date>
+  <parameters>
+    </parameters>
+  <actions>
+    <action>
+      <name>Start</name>
+      <description/>
+      <type>SPECIAL</type>
+      <attributes/>
+      <DayOfMonth>1</DayOfMonth>
+      <hour>12</hour>
+      <intervalMinutes>60</intervalMinutes>
+      <intervalSeconds>0</intervalSeconds>
+      <minutes>0</minutes>
+      <repeat>N</repeat>
+      <schedulerType>0</schedulerType>
+      <weekDay>1</weekDay>
+      <parallel>N</parallel>
+      <xloc>50</xloc>
+      <yloc>50</yloc>
+      <attributes_hac/>
+    </action>
+    <action>
+      <name>Run check files locked on server</name>
+      <description/>
+      <type>WORKFLOW</type>
+      <attributes/>
+      <run_configuration>remote</run_configuration>
+      <filename>${PROJECT_HOME}/0010-check-files-locked.hwf</filename>
+      <params_from_previous>N</params_from_previous>
+      <exec_per_row>N</exec_per_row>
+      <set_logfile>N</set_logfile>
+      <logfile/>
+      <logext/>
+      <add_date>N</add_date>
+      <add_time>N</add_time>
+      <loglevel>Basic</loglevel>
+      <wait_until_finished>Y</wait_until_finished>
+      <create_parent_folder>N</create_parent_folder>
+      <parameters>
+        <pass_all_parameters>Y</pass_all_parameters>
+      </parameters>
+      <set_append_logfile>N</set_append_logfile>
+      <parallel>N</parallel>
+      <xloc>250</xloc>
+      <yloc>50</yloc>
+      <attributes_hac/>
+    </action>
+  </actions>
+  <hops>
+    <hop>
+      <from>Start</from>
+      <to>Run check files locked on server</to>
+      <enabled>Y</enabled>
+      <evaluation>Y</evaluation>
+      <unconditional>Y</unconditional>
+    </hop>
+  </hops>
+  <notepads>
+  </notepads>
+  <attributes/>
+</workflow>

--- a/plugins/actions/checkfilelocked/src/main/java/org/apache/hop/workflow/actions/checkfilelocked/ActionCheckFilesLocked.java
+++ b/plugins/actions/checkfilelocked/src/main/java/org/apache/hop/workflow/actions/checkfilelocked/ActionCheckFilesLocked.java
@@ -75,11 +75,12 @@ public class ActionCheckFilesLocked extends ActionBase implements Cloneable, IAc
     return new ActionCheckFilesLocked(this);
   }
 
-  public ActionCheckFilesLocked(ActionCheckFilesLocked a) {
-    this();
-    this.argFromPrevious = a.argFromPrevious;
-    this.includeSubfolders = a.includeSubfolders;
-    for (CheckedFile checkedFile : a.checkedFiles) {
+  public ActionCheckFilesLocked(ActionCheckFilesLocked other) {
+    super(other.getName(), other.getDescription(), other.getPluginId());
+    this.checkedFiles = new ArrayList<>();
+    this.argFromPrevious = other.argFromPrevious;
+    this.includeSubfolders = other.includeSubfolders;
+    for (CheckedFile checkedFile : other.checkedFiles) {
       this.checkedFiles.add(new CheckedFile(checkedFile));
     }
   }

--- a/plugins/actions/checkfilelocked/src/test/java/org/apache/hop/workflow/actions/checkfilelocked/ActionCheckFilesLockedTest.java
+++ b/plugins/actions/checkfilelocked/src/test/java/org/apache/hop/workflow/actions/checkfilelocked/ActionCheckFilesLockedTest.java
@@ -43,4 +43,27 @@ class ActionCheckFilesLockedTest {
     Assertions.assertEquals("/tmp/folder2", action.getCheckedFiles().get(1).getName());
     Assertions.assertEquals(".*\\.zip$", action.getCheckedFiles().get(1).getWildcard());
   }
+
+  /**
+   * Resource export clones every action before serializing it. A clone that loses its name or
+   * plugin id ends up in the export with an empty name and without its type, and the exported
+   * workflow then fails to load on a remote server. See issue #5644.
+   */
+  @Test
+  void cloneKeepsNamePluginIdAndFields() {
+    ActionCheckFilesLocked action = new ActionCheckFilesLocked("Check files locked");
+    action.setDescription("a description");
+    action.setPluginId("CHECK_FILES_LOCKED");
+    action.setIncludeSubfolders(true);
+    action.getCheckedFiles().add(new ActionCheckFilesLocked.CheckedFile("/tmp/file", ".*\\.txt$"));
+
+    ActionCheckFilesLocked copy = (ActionCheckFilesLocked) action.clone();
+
+    Assertions.assertEquals("Check files locked", copy.getName());
+    Assertions.assertEquals("a description", copy.getDescription());
+    Assertions.assertEquals("CHECK_FILES_LOCKED", copy.getPluginId());
+    Assertions.assertTrue(copy.isIncludeSubfolders());
+    Assertions.assertEquals(1, copy.getCheckedFiles().size());
+    Assertions.assertEquals("/tmp/file", copy.getCheckedFiles().get(0).getName());
+  }
 }

--- a/plugins/actions/ftp/src/main/java/org/apache/hop/workflow/actions/ftpput/ActionFtpPut.java
+++ b/plugins/actions/ftp/src/main/java/org/apache/hop/workflow/actions/ftpput/ActionFtpPut.java
@@ -163,7 +163,7 @@ public class ActionFtpPut extends ActionBase implements Cloneable, IAction, IFtp
   }
 
   public ActionFtpPut(ActionFtpPut a) {
-    super(a.getName(), a.getDescription());
+    super(a.getName(), a.getDescription(), a.getPluginId());
     this.serverName = a.serverName;
     this.serverPort = a.serverPort;
     this.userName = a.userName;

--- a/plugins/actions/ftp/src/test/java/org/apache/hop/workflow/actions/ftpput/ActionFtpPutTest.java
+++ b/plugins/actions/ftp/src/test/java/org/apache/hop/workflow/actions/ftpput/ActionFtpPutTest.java
@@ -69,4 +69,25 @@ class ActionFtpPutTest {
     assertEquals("proxy-username", action.getSocksProxyUsername());
     assertEquals("proxy-password", action.getSocksProxyPassword());
   }
+
+  /**
+   * Resource export clones every action. A clone that loses its plugin id is written to the export
+   * without its type element, and the exported workflow then fails to load on a remote server. See
+   * issue #5644.
+   */
+  @Test
+  void cloneKeepsNameDescriptionAndPluginId() {
+    ActionFtpPut action = new ActionFtpPut();
+    action.setName("Put files");
+    action.setDescription("a description");
+    action.setPluginId("FTP_PUT");
+    action.setServerName("ftp-server");
+
+    ActionFtpPut copy = (ActionFtpPut) action.clone();
+
+    assertEquals("Put files", copy.getName());
+    assertEquals("a description", copy.getDescription());
+    assertEquals("FTP_PUT", copy.getPluginId());
+    assertEquals("ftp-server", copy.getServerName());
+  }
 }

